R/cross.R

```diff
@@ -0,0 +1,160 @@
+#' Produce all combinations of list elements
+#'
+#' \code{cross()} returns the product set of the elements of \code{.x}
+#' and \code{.y}. It is similar to \code{expand.grid()} but it returns
+#' a list. By default, the cartesian product is returned in wide
+#' format. This makes it more amenable to mapping
+#' operations. Switching \code{.wide} to \code{FALSE} turns the output
+#' to the long format, the equivalent to \code{expand.grid()}'s
+#' outputs.
+#'
+#' \code{cross3()} takes three three arguments instead of two and
+#' returns the cartesian product of the elements of the three
+#' objects. \code{cross_n()} takes a list \code{.l} and returns the
+#' cartesian product of all its elements. If \code{.l} is a data
+#' frame, \code{cross_n()} returns a data frame.
+#'
+#' When the number of combinations is large and the individual elements
+#' are heavy memory-wise, it is often useful to filter unwanted
+#' combinations on the fly. \code{.filter} must be a predicate
+#' functio! that takes the same number of arguments as the number of
+#' crossed objects (2 for \code{cross()}, 3 for \code{cross3()},
+#' \code{length(.l)} for \code{cross_n()}) and returns \code{TRUE} or
+#' \code{FALSE}. The combinations where the predicate function returns
+#' \code{TRUE} will be removed from the result.
+#'
+#' @seealso expand.grid
+#' @param .x,.y,.z Lists or atomic vectors.
+#' @param .l A list of lists or atomic vectors.
+#' @param .wide If \code{TRUE}, returns a list of unique combinations.
```

Comment:
The more I think about this the more I like it. It would make `cross()` a bit different to other functions, but it would be more likely to generate the correct output by default.